# Vector search for SOC

### Initialize connection to elasticsearch.

In [2]:
from dotenv import load_dotenv
from elasticsearch import Elasticsearch
import os
from pandas import json_normalize
import pandas as pd
import json
from IPython.display import display


pd.set_option('display.max_colwidth', None)

load_dotenv()
cloud_id = os.getenv('cloud_id')
api_key = os.getenv('api_key')


client = Elasticsearch(
    cloud_id=cloud_id,
    api_key=api_key
)


### 1. Start by searching for suspicious access to the S3 bucket with secret file
    First we can see example of problem with approximate kNN search capturing results that go in the right direction but lack precision. Model used here is SecureBERT. After experimenting wiht it I switched to elasticsearch sparse vector model called ELSER.

In [16]:
query = "get object from s3 bucket name bsides-zagreb"

''' Some other query examples also tried 
    query = "bucket name bsides-zagreb data exfiltration to external IP address"
    query = "bucket bsides-zagreb data leak"
    query = "get data from s3 bucket named bsides-zagreb"
''' 


response = client.search(
    index=".ds-logs-aws.cloudtrail-default-2025.03.04-000014",
    size=20,
    query={
      "knn": {
        "field": "message_secure_bert_max_inp",
        "k": 20,
        "num_candidates": 10000,
        "query_vector_builder": {
          "text_embedding": {
            "model_id": "secure-bert",
            "model_text": f"{query}"
      }
    }
  }
},
)

results = []

for hit in response["hits"]["hits"]:
    
    event_original = json.loads(hit["_source"]["event"]["original"])
    event_original['score'] = hit["_score"]
    results.append(event_original)
        
df_all = pd.DataFrame(results)
df_cleaned= df_all[['score','eventTime','userIdentity', 'eventName', 'awsRegion', "sourceIPAddress", 'userAgent', 'requestParameters', 'resources']]

display(df_cleaned)


score             eventTime  \
0   3.863222  2025-02-25T09:17:32Z   
1   3.858194  2025-02-25T09:17:32Z   
2   3.854832  2025-02-25T09:17:02Z   
3   3.852299  2025-02-25T17:43:43Z   
4   3.852140  2025-02-25T19:43:42Z   
5   3.851715  2025-02-26T11:14:02Z   
6   3.851162  2025-02-26T10:45:44Z   
7   3.851126  2025-02-25T16:44:45Z   
8   3.850445  2025-02-25T15:43:46Z   
9   3.850220  2025-02-25T21:44:46Z   
10  3.849691  2025-02-25T15:45:24Z   
11  3.849667  2025-02-25T17:42:43Z   
12  3.849281  2025-02-25T23:45:00Z   
13  3.848345  2025-02-25T10:42:36Z   
14  3.848326  2025-02-26T05:44:08Z   
15  3.847971  2025-02-25T23:43:49Z   
16  3.846647  2025-02-25T23:42:40Z   
17  3.846571  2025-02-26T10:43:46Z   
18  3.846460  2025-02-25T23:44:49Z   
19  3.846456  2025-02-26T11:40:24Z   

                                                                                                                                                                                                                                 userIdentity  \
0   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
1   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
2   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
3   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
4   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
5   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
6   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
7   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
8   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
9   {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
10  {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
11  {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn:aws:iam::538539780105:user/elastic-agent-cloudtrail', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AEZOIJ7U72', 'userName': 'elastic-agent-cloudtrail'}   
12  {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE6C35PLXLV', 'arn': 'arn

#### Here we can see the same query with sparse vector search using Elastics ELSER model. We get exact results ranked at top so we will continue with sparse vector search.

In [18]:
query = "get object from s3 bucket name bsides-zagreb"

#query = "download object from s3 bucket from public IP" results in top 5

response = client.search(
    index=".ds-logs-aws.cloudtrail-default-2025.03.04-000016",
    size=5,
    query={
      "sparse_vector": {
        "field": "message_elser_v2",
         "inference_id": ".elser_model_2_linux-x86_64",
          "query": query

  }
},
)

results = []

for hit in response["hits"]["hits"]:
    
    event_original = json.loads(hit["_source"]["event"]["original"])
    event_original['score'] = hit["_score"]
    results.append(event_original)

df_all = pd.DataFrame(results)
df_cleaned= df_all[['score','eventTime','userIdentity', 'eventName', 'awsRegion', "sourceIPAddress", 'userAgent', 'requestParameters', 'resources']]


display(df_cleaned)

score             eventTime  \
0  20.751670  2025-03-03T15:16:30Z   
1  20.287140  2025-03-03T15:16:30Z   
2  19.995228  2025-03-03T15:13:31Z   
3  19.560778  2025-03-03T13:27:42Z   
4  18.782330  2025-03-03T13:48:47Z   

                                                                                                                                                                                                                                                           userIdentity  \
0                                                 {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE7KVE2CM3O', 'arn': 'arn:aws:iam::538539780105:user/lambda_s3_user', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AESBKH7OGC', 'userName': 'lambda_s3_user'}   
1                                                 {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE7KVE2CM3O', 'arn': 'arn:aws:iam::538539780105:user/lambda_s3_user', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AESBKH7OGC', 'userName': 'lambda_s3_user'}   
2                                                 {'type': 'IAMUser', 'principalId': 'AIDAX2Y3Y5AE7KVE2CM3O', 'arn': 'arn:aws:iam::538539780105:user/lambda_s3_user', 'accountId': '538539780105', 'accessKeyId': 'AKIAX2Y3Y5AESBKH7OGC', 'userName': 'lambda_s3_user'}   
3  {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AERI7TIC4B', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
4  {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AE6TDYY4HN', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   

                          eventName     awsRegion sourceIPAddress  \
0                         GetObject  eu-central-1    95.168.107.4   
1                        HeadObject  eu-central-1    95.168.107.4   
2                       ListObjects  eu-central-1    95.168.107.4   
3  GetBucketObjectLockConfiguration  eu-central-1    95.168.107.4   
4  GetBucketObjectLockConfiguration  eu-central-1    95.168.107.4   

                                                                                                                                                             userAgent  \
0  [aws-cli/1.36.17 md/Botocore#1.35.71 ua/2.0 os/linux#5.15.0-86-generic md/arch#x86_64 lang/python#3.10.12 md/pyimpl#CPython cfg/retry-mode#legacy botocore/1.35.71]   
1  [aws-cli/1.36.17 md/Botocore#1.35.71 ua/2.0 os/linux#5.15.0-86-generic md/arch#x86_64 lang/python#3.10.12 md/pyimpl#CPython cfg/retry-mode#legacy botocore/1.35.71]   
2  [aws-cli/1.36.17 md/Botocore#1.35.71 ua/2.0 os/linux#5.15.0-86-generic md/arch#x86_64 lang/python#3.10.12 md/pyimpl#CPython cfg/retry-mode#legacy botocore/1.35.71]   
3                                                              [Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36]   
4                                                              [Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36]   

                                                                                                                                                  requestParameters  \
0                                       {'bucketName': 'bsides-zagreb', 'Host': 'bsides-zagreb.s3.eu-central-1.amazonaws.com', 'key': 'super_secret_text_file.txt'}   
1                                       {'bucketName': 'bsides-zagreb', 'Host': 'bsides-zagreb.s3.eu-central-1.amazonaws.com', 'key': 'super_secret_text_file.txt'}   
2  {'list-type': '2', 'bucketName': 'bsides-zagreb', 'encoding-type': 'url', 'prefix': '', 'delimiter': '/', 'Host': 'bsides-zagreb.s3.eu-central-1.amazonaws.com'}   
3                                                                      

### 2. See what else we can find related to before mentioned user **lambda_s3_user**
- By user name we can't really find connection to previous actions, this indicates that attacker escalated to this user.
- We still have IP address that commands came from. From username we can assume that is is somehow related to lambda functions so we can check the actions performed on any lambda function.
- Lambda functions can often leak hardcoded credentials in code or env variables.
- Here we can see some problems with getting high score results for get function query since the event name for it is actually GetFunction20150331v2.
- Listfunction api calls are executed via CLI using assumed role **S3_Lambda**

In [458]:
query = "listfunctions"

#query = "actions by user name lambda_s3_user from aws cli"


response = client.search(
    index=".ds-logs-aws.cloudtrail-default-2025.03.04-000016",
    size=15,
    query={
      "sparse_vector": {
        "field": "message_elser_v2",
         "inference_id": ".elser_model_2_linux-x86_64_ingest",
          "query": query

  }
},
)

results = []

for hit in response["hits"]["hits"]:
    
    event_original = json.loads(hit["_source"]["event"]["original"])
    event_original['score'] = hit["_score"]
    results.append(event_original)


df_all = pd.DataFrame(results)
df_cleaned = df_all[['score','eventTime','userIdentity', 'eventName', 'awsRegion', "sourceIPAddress", 'userAgent', 'requestParameters']]

pd.set_option('display.max_colwidth', None)

display(df_cleaned)

score             eventTime  \
0   11.164793  2025-03-03T15:08:55Z   
1   10.742123  2025-03-03T13:52:10Z   
2   10.726970  2025-03-03T13:53:41Z   
3   10.449755  2025-03-03T13:54:31Z   
4   10.370233  2025-03-03T13:55:25Z   
5    9.990437  2025-03-03T13:56:32Z   
6    9.940550  2025-03-03T15:07:29Z   
7    9.581711  2025-03-03T13:53:43Z   
8    9.279804  2025-03-03T13:52:15Z   
9    9.200465  2025-03-03T13:54:31Z   
10   9.099659  2025-03-03T13:55:29Z   
11   8.997975  2025-03-03T13:54:36Z   
12   8.814835  2025-03-03T13:54:37Z   
13   8.748467  2025-03-03T14:19:05Z   
14   8.708790  2025-03-03T13:55:29Z   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     userIdentity  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             {'type': 'AWSService', 'invokedBy': 'AWS Internal'}   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             {'type': 'AWSService', 'invokedBy': 'AWS Internal'}   
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             {'type': 'AWSService', 'invokedBy': 'AWS Internal'}   
3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             {'type': 'AWSService', 'invokedBy': 'AWS Internal'}   
4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             {'type': 'AWSService', 'invokedBy': 'AWS Internal'}   
5                                                                                        

## 4. Investigate events related to instance i-0905d72baf3abed14
    From role ARN arn:aws:sts::538539780105:assumed-role/S3_Lambda/*i-0905d72baf3abed14* we can conclude that this was role attached to an EC2 instance with ID i-0905d72baf3abed14. Requests are coming from public IP and not directly from instance which leads us to assuming temporary credentials related to this EC2 instance have been stolen.

In [4]:
query = "instance metadata"


response = client.search(
    index="logs-aws.cloudtrail*",
    size=15,
    query={
      "sparse_vector": {
        "field": "message_elser_v2",
         "inference_id": ".elser_model_2_linux-x86_64_ingest",
          "query": query

  }
},
)

results = []

for hit in response["hits"]["hits"]:

    event = json.loads(hit["_source"]["event"]["original"])

        
    event["score"] = float(hit["_score"])
    results.append(event)

df_all = pd.DataFrame(results)
df_cleaned= df_all[['score','eventTime','userIdentity', 'eventName', 'awsRegion', "sourceIPAddress", 'userAgent', 'requestParameters']]

pd.set_option('display.max_colwidth', None)

display(df_cleaned)

score             eventTime  \
0   8.260045  2025-03-03T13:02:47Z   
1   8.242740  2025-03-03T13:03:01Z   
2   8.188886  2025-03-03T11:02:41Z   
3   8.184713  2025-03-03T12:46:08Z   
4   8.175696  2025-03-03T13:00:15Z   
5   8.114229  2025-03-03T13:01:31Z   
6   8.106265  2025-03-03T09:48:50Z   
7   8.103455  2025-03-03T13:00:12Z   
8   8.095818  2025-03-03T08:17:25Z   
9   7.992964  2025-03-03T11:33:37Z   
10  7.974815  2025-03-03T13:03:49Z   
11  7.964776  2025-03-03T12:46:08Z   
12  7.960899  2025-03-03T13:00:13Z   
13  7.947421  2025-03-03T10:57:34Z   
14  7.920418  2025-03-03T13:02:48Z   

                                                                                                                                                                                                                                                                                                            userIdentity  \
0                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AETZBV3KQ3', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
1                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AETZBV3KQ3', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
2   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEWA6423XI', 'sessionContext': {'sessionIssuer': {}, 'webIdFederationData': {}, 'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
3                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEQPTQTSZJ', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
4                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AETZBV3KQ3', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
5                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AESX7XQOGC', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
6                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AE52RL4LVH', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
7                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AETZBV3KQ3', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
8                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEZAGZGTGE', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
9                                                   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AE4TVWXVZS', 'sessionContext': {'attribute

## 5. Pivot to Application load balancer logs for more info

- Cloudtrail by itself doesn't log requests to instance metadata service
- We can connect API explitation attempts to the compromised instance
- From ALB logs we can confirm that attacker abused instance metadata service to obtain credentials to impersonate role S3_lambda.

In [12]:
query = "meta data iam S3_lambda"


response = client.search(
    index="logs-aws.elb_logs-default",
    size=15,
    query={
      "sparse_vector": {
        "field": "message_elser_v2",
         "inference_id": ".elser_model_2_linux-x86_64_ingest",
          "query": query
  }
},
)

for hit in response["hits"]["hits"]:
        
    print(f"Score:{hit['_score']}")
    print(f"Message: {hit['_source']['event']['original']}",end=f"\n\n{120*'-'}\n\n")
    

Score:9.403352
Message: http 2025-03-03T14:52:17.471449Z app/vuln-app-alb/72fb0c71304e1795 95.168.107.4:29916 10.0.0.78:9000 0.001 0.006 0.000 200 200 361 1817 "GET http://vuln-app-alb-404212911.eu-central-1.elb.amazonaws.com:8234/get?url=http%3A%2F%2F169.254.169.254%2Flatest%2Fmeta-data%2Fiam%2Fsecurity-credentials%2FS3_Lambda HTTP/1.1" "HTTPie/3.2.4" - - arn:aws:elasticloadbalancing:eu-central-1:538539780105:targetgroup/vuln-app-targets/28bc56e20e82ee3f "Root=1-67c5c221-5e9ceb4670872dca2cfc87fa" "-" "-" 0 2025-03-03T14:52:17.464000Z "forward" "-" "-" "10.0.0.78:9000" "200" "-" "-" TID_d29df88a9bdf6747aa8fdc8672cd8e73

------------------------------------------------------------------------------------------------------------------------

Score:4.930862
Message: http 2025-03-03T14:27:42.122817Z app/vuln-app-alb/72fb0c71304e1795 95.168.107.4:33746 10.0.0.78:9000 0.001 0.002 0.000 405 405 320 208 "GET http://vuln-app-alb-404212911.eu-central-1.elb.amazonaws.com:8234/put?url=http%3A%2F%2

### 6. Find more about the connection with specific instance by querying for target group
    If we can find event of target group creation and registration we can confirm that instance belongs to vulnerable target group and was the entry point for the attacker.

In [20]:
query = "target group vuln-app-targets i-0905d72baf3abed14"


response = client.search(
    index="logs-aws.cloudtrail*",
    size=15,
    query={
      "sparse_vector": {
        "field": "message_elser_v2",
         "inference_id": ".elser_model_2_linux-x86_64",
          "query": query

  }
},
)

results = []

for hit in response["hits"]["hits"]:

    event = json.loads(hit["_source"]["event"]["original"])

        
    event["score"] = float(hit["_score"])
    results.append(event)  

df_all = pd.DataFrame(results)
df_cleaned= df_all[['score','eventTime','userIdentity', 'eventName', 'awsRegion', "sourceIPAddress", 'userAgent', 'requestParameters']]


pd.set_option('display.max_colwidth', None)

display(df_cleaned)

score             eventTime  \
0   18.524036  2025-03-03T07:56:13Z   
1   18.341840  2025-03-03T12:52:55Z   
2   18.207516  2025-03-03T11:33:36Z   
3   18.010761  2025-03-03T12:56:24Z   
4   17.958190  2025-03-03T12:59:21Z   
5   17.755571  2025-03-03T07:53:14Z   
6   17.286861  2025-03-03T12:57:36Z   
7   17.251030  2025-03-03T12:59:42Z   
8   17.240840  2025-03-03T12:49:36Z   
9   17.203327  2025-03-03T11:39:26Z   
10  17.141563  2025-03-03T12:47:32Z   
11  17.098251  2025-03-03T11:37:59Z   
12  17.096430  2025-03-03T08:17:50Z   
13  17.079070  2025-03-03T12:47:28Z   
14  16.969324  2025-03-03T12:51:00Z   

                                                                                                                                                                                                                                                            userIdentity  \
0   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEWYXOCTBN', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
1   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEZI24OWWS', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
2   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AE3RJTLRG4', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
3   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEVIVCRLH7', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
4   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEVIVCRLH7', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
5   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEWYXOCTBN', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
6   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEZI24OWWS', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
7   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AEVIVCRLH7', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
8   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AERUJIFQTM', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
9   {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AE3RJTLRG4', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
10  {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AERUJIFQTM', 'sessionContext': {'attributes': {'creationDate': '2025-03-03T07:49:19Z', 'mfaAuthenticated': 'true'}}}   
11  {'type': 'Root', 'principalId': '538539780105', 'arn': 'arn:aws:iam::538539780105:root', 'accountId': '538539780105', 'accessKeyId': 'ASIAX2Y3Y5AE3RJTLRG4', 'sessi